File: tasks/translation/output/Qwen/QwQ-32B-AWQ/codenet/direct/javadoc/uncomment/direct/Java/Go/s584572995_commented.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

const mod = 1e9 + 7

func modpow(x, y int) int {
	res := 1
	x %= mod
	for y > 0 {
		if y%2 != 0 {
			res = res * x % mod
		}
		x = x * x % mod
		y /= 2
	}
	return res
}

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	scanner.Split(bufio.ScanWords)
	var tokens []string
	for scanner.Scan() {
		tokens = append(tokens, scanner.Text())
	}
	N, _ := strconv.Atoi(tokens[0])
	to := make([][]int, N)
	edges := make([][2]int, N-1)
	idx := 1
	for i := 0; i < N-1; i++ {
		aStr := tokens[idx]
		bStr := tokens[idx+1]
		A, _ := strconv.Atoi(aStr)
		B, _ := strconv.Atoi(bStr)
		A--
		B--
		edges[i] = [2]int{A, B}
		to[A] = append(to[A], B)
		to[B] = append(to[B], A)
		idx += 2
	}

	dp := make([]int, N)
	var dfs func(v, parent int) int
	dfs = func(v, parent int) int {
		res := 1
		for _, next := range to[v] {
			if next == parent {
				continue
			}
			res += dfs(next, v)
		}
		dp[v] = res
		return res
	}
	dfs(0, -1)

	var mulsum int
	for _, e := range edges {
		a, b := e[0], e[1]
		child := a
		if dp[a] > dp[b] {
			child = b
		}
		aSize := dp[child]
		bSize := N - aSize
		aTerm := (modpow(2, aSize) - 1) % mod
		bTerm := (modpow(2, bSize) - 1) % mod
		term := aTerm * bTerm % mod
		mulsum = (mulsum + term) % mod
	}

	cases := modpow(2, N)
	PN := (mulsum + (cases-1+mod)%mod) % mod
	BN := N * modpow(2, N-1) % mod
	WN := (PN - BN + mod) % mod
	invCases := modpow(cases, mod-2)
	ans := WN * invCases % mod
	fmt.Println(ans)
}

// End of Code

File: tasks/translation/output/Qwen/QwQ-32B-AWQ/codenet/direct/javadoc/uncomment/direct/Java/Python/s584572995_commented.py
Here's the translated Python code:


import sys

MOD = 10**9 + 7

def main():
    sys.setrecursionlimit(1 << 25)
    N = int(sys.stdin.readline())
    to = [[] for _ in range(N)]
    edges = []
    for _ in range(N-1):
        A,B = map(int, sys.stdin.readline().split())
        A -= 1
        B -= 1
        to[A].append(B)
        to[B].append(A)
        edges.append((B, A))
    
    dp = [0] * N
    stack = [(0, -1, False)]
    while stack:
        node, parent, visited = stack.pop()
        if not visited:
            stack.append((node, parent, True))
            for neighbor in to[node]:
                if neighbor != parent:
                    stack.append((neighbor, node, False))
        else:
            dp[node] = 1
            for neighbor in to[node]:
                if neighbor != parent:
                    dp[node] += dp[neighbor]
    
    mulsum = 0
    for e in edges:
        B, A = e
        c1, c2 = B, A
        if dp[c1] > dp[c2]:
            child = c2
        else:
            child = c1
        a = dp[child]
        b = N - a
        term = (pow(2, a, MOD) - 1) * (pow(2, b, MOD) - 1) % MOD
        mulsum = (mulsum + term) % MOD
    
    cases = pow(2, N, MOD)
    PN = (mulsum + (cases - 1)) % MOD
    BN = (N * pow(2, N-1, MOD)) % MOD
    WN = (PN - BN) % MOD
    if WN < 0:
        WN += MOD
    inv_cases = pow(cases, MOD-2, MOD)
    ans = WN * inv_cases % MOD
    print(ans)

if __name__ == "__main__":
    main()
# End of Code
